components/places/src/import/ios_bookmarks.rs

```diff
@@ -0,0 +1,566 @@
+/// history of iOS bookmark sync (accurate last modified times, for example).
+///
+/// As a result, the goals of this import are as follows:
+///
+/// 1. Any locally created items must be persisted.
+///
+/// 2. Any items from remote machines that are visible to the user must be
+///    persisted. (Note: before writing this, most of us believed that iOS wiped
+///    it's view of remote bookmarks on sync sign-out. Apparently it does not,
+///    and it's unclear if it ever did).
+///
+/// Additionally, it's worth noting that we assume that the iOS tree is
+/// relatively well-formed. We do leverage `dogear` for the merge, to avoid
+/// anything absurd, but for the most part the validation is fairly loose. If we
+/// see anything we don't like (URL we don't allow, for example), we skip it.
+///
+/// ### Unsupported features
+///
+/// As such, the following things are explicitly not imported:
+///
+/// - Livemarks: We don't support them in our database anyway.
+/// - Tombstones: This shouldn't matter for non-sync users.
+/// - Queries: Not displayed or creatable in iOS UI, and only half-supported in
+///   this database.
+///
+/// Some of this (queries, really) is a little unfortunate, since it's
+/// theoretically possible for someone to care, but this should only happen for
+/// users:
+///
+/// - Who once used sync, but no longer do.
+/// - Who used this feature when they used sync.
+/// - Who no longer have access to any firefoxes from when they were sync users,
+///   other than this iOS device.
+///
+/// For these users, upon signing into sync once again, they will lose the data
+/// in question.
+///
+/// ### Basic process
+///
+/// - Attach the iOS database.
+/// - Slurp records into a temp table "iosBookmarksStaging" from iOS database.
+/// - Add any entries to moz_places that are needed (in practice, they'll all be
+///   needed, we don't yet store history for iOS).
+/// - Fill mirror using iosBookmarksStaging.
+/// - Fill mirror with tags from iosBookmarksStaging.
+/// - Fill mirror structure using both iOS database and iosBookmarksStaging.
+/// - Run dogear merge
+/// - Use iosBookmarksStaging to fixup the data that was actually inserted.
+/// - Update frecency for new items.
+/// - Cleanup (Delete mirror and mirror structure, detach iOS database, etc).
+pub fn import_ios_bookmarks(
+    places_api: &PlacesApi,
+    path: impl AsRef<std::path::Path>,
+) -> Result<()> {
+    let url = crate::util::ensure_url_path(path)?;
+    do_import_ios_bookmarks(places_api, url)
+}
+
+fn do_import_ios_bookmarks(places_api: &PlacesApi, ios_db_file_url: Url) -> Result<()> {
+    let conn = places_api.open_sync_connection()?;
+
+    let scope = conn.begin_interrupt_scope();
+
+    sql_fns::define_functions(&conn)?;
+
+    // Not sure why, but apparently beginning a transaction sometimes
+    // fails if we open the DB as read-only. Hopefully we don't
+    // unintentionally write to it anywhere...
```

Comment:
Oh, interesting! I was wondering if that's because starting a transaction locks all databases for writing, even if you're only reading from one—but that's not right. 🤷‍♀

Reply:
Yeah, it's subtle, it also didn't happen 100% of the time.